plugins/fileservices/fileservices.cpp

```diff
@@ -198,63 +198,136 @@ static IPropertyTree *getEnvironmentTree(IConstEnvironment * daliEnv)
     return getHPCCEnvironment();
 }
 
-static StringAttr espurl;           // Default ESP url if none specified
+static void setServerAccess(CClientFileSpray &server, IConstWorkUnit * wu)
+{
+    StringBuffer user, password, token;
+    wu->getSecurityToken(StringBufferAdaptor(token));
+    extractToken(token.str(), wu->queryWuid(), StringBufferAdaptor(user), StringBufferAdaptor(password));
+    server.setUsernameToken(user.str(), password.str(), "");
+}
+
+typedef MapStringTo<time_t> StrMapToEpoch;
+static StrMapToEpoch espUrlsLastHeartBeat;
 static CriticalSection espURLcrit;
 
+static void setWsFSUrlAliveTimeStamp(const char * url, time_t aliveTime)
+{
+    StringBuffer key;
+    CriticalBlock b(espURLcrit);
+    espUrlsLastHeartBeat.setValue(key.set(url).str(), aliveTime);
```

Comment:
I don't understand this answer. The copy into key does not make any sense to me unless you are trying to translate nullptr to "" (but it's an inefficient way to do that...)